File: day9/solve-2.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

func moveBlock(i int, blocks_index []int, block_size []int, spaces map[int]int) {
	is_movable := false
	old_i := -1
	old_s := 1000
	for i_s, s_s := range spaces {
		if i_s < blocks_index[i] {
			if s_s >= block_size[i] {
				blocks_index[i] = i_s
				is_movable = true
				old_i = i_s
				old_s = s_s
				break
			}
		}
	}
	if is_movable {
		spaces[old_i+block_size[i]] = old_s - block_size[i]
		spaces[old_i] = -1
	}
}

func solve2() {
	file, err := os.Open("in")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var line string
	if scanner.Scan() {
		line = scanner.Text()
	}

	block_size := []int{}
	blocks_index := []int{}
	spaces := make(map[int]int)
	index := 0
	for i, s := range line {
		s = s - '0'
		if i%2 == 0 {
			block_size = append(block_size, int(s))
			blocks_index = append(blocks_index, index)
			for k := 0; k < int(s); k++ {
				index++
			}
		} else {
			spaces[index] = int(s)
			for k := 0; k < int(s); k++ {
				index++
			}
		}
	}

	for i := len(block_size) - 1; i >= 0; i-- {
		moveBlock(i, blocks_index, block_size, spaces)
	}

	sum := 0
	for b := 1; b < len(block_size); b++ {
		for b_i := blocks_index[b]; b_i < blocks_index[b]+block_size[b]; b_i++ {
			sum += b * b_i
		}
	}
	fmt.Println(sum)
}

File: day9/cpp/solve-2.cpp
#include <cstdint>
#include <iostream>
#include <fstream>
#include <vector>
#include <map>
#include <string>

using namespace std;

void moveBlock(int i, vector<int>& blocks_index, vector<int>& block_size, map<int, int>& spaces) {
    bool isMovable = false;
    int oldIndex = -1;
    int oldSize = 1000;

    for (auto& [i_s, s_s] : spaces) {
        if (i_s < blocks_index[i]) {
            if (s_s >= block_size[i]) {
                blocks_index[i] = i_s;
                isMovable = true;
                oldIndex = i_s;
                oldSize = s_s;
                break;
            }
        }
    }

    if (isMovable) {
        spaces[oldIndex + block_size[i]] = oldSize - block_size[i];
        spaces[oldIndex] = -1;
    }
}

void solve2() {
    ifstream file("in");
    if (!file.is_open()) {
        cerr << "Error: Unable to open file\n";
        return;
    }

    string line;
    getline(file, line);

    vector<int> block_size;
    vector<int> blocks_index;
    map<int, int> spaces;

    int index = 0;
    for (size_t i = 0; i < line.size(); ++i) {
        int s = line[i] - '0';
        if (i % 2 == 0) {
            block_size.push_back(s);
            blocks_index.push_back(index);
            index += s;
        } else {
            spaces[index] = s;
            index += s;
        }
    }

    for (int i = block_size.size() - 1; i >= 0; --i) {
        moveBlock(i, blocks_index, block_size, spaces);
    }

    uint64_t sum = 0;
    for (size_t b = 1; b < block_size.size(); ++b) {
        for (int b_i = blocks_index[b]; b_i < blocks_index[b] + block_size[b]; ++b_i) {
            sum += b * b_i;
        }
    }
    cout << sum << endl;
}

int main() {
    solve2();
    return 0;
}

